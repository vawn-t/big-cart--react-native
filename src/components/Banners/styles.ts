import { StyleSheet } from 'react-native';
import { colors } from '@themes/index';

const styles = StyleSheet.create({
  container: {
    height: 283,
  },
  indicatorWrapper: {
    position: 'absolute',
    flexDirection: 'row',
    columnGap: 5,
    bottom: 30,
    left: 15,
  },
  indicator: {
    backgroundColor: colors.primary.light,
    borderRadius: 4,
    width: 8,
    height: 8,
  },

  bannerContainer: {
    flex: 1,
    width: '100%',
    paddingHorizontal: 45,
    paddingTop: 45,
    justifyContent: 'center',
  },
  text: { width: 175 },
});

export default styles;
