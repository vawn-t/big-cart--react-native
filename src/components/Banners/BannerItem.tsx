import { ImageBackground } from 'react-native';

// Models
import { Slide } from '@models/index';

// Components
import { Typography } from '@components/common';

// Types
import { FontWeight, SizeType, TypoVariant } from '@interfaces/index';

// Styles
import styles from './styles';

interface IProps {
  item: Slide;
  width: number;
}

const BannerItem = ({ item, width }: IProps) => (
  <ImageBackground
    source={item.image}
    style={[styles.bannerContainer, { width }]}
    resizeMode="cover"
  >
    <Typography
      text={item.title}
      style={styles.text}
      size={SizeType.Large}
      variant={TypoVariant.Paragraph2}
      fontWeight={FontWeight.SemiBold}
    />
  </ImageBackground>
);

export default BannerItem;
