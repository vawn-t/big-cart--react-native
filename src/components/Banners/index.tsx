import { useCallback, useRef, useState } from 'react';
import {
  FlatList,
  LayoutChangeEvent,
  View,
  ViewToken,
  ViewabilityConfigCallbackPairs,
} from 'react-native';

// Components
import BannerItem from './BannerItem';

// Themes
import { colors } from '@themes/index';

// Mocks
import { homeSlides } from '@mocks/index';

// Styles
import styles from './styles';

const Banners = () => {
  const [layout, setLayout] = useState<{ width: number }>({ width: 0 });
  const [currentActiveSlide, setCurrentActiveSlide] = useState(0);

  const viewabilityConfig = {
    viewAreaCoveragePercentThreshold: 100,
    waitForInteraction: true,
  };

  const onViewableItemsChanged = useCallback(
    ({ changed }: { changed: Array<ViewToken> }) => {
      if (
        changed &&
        changed.length > 0 &&
        typeof changed[0].index === 'number'
      ) {
        setCurrentActiveSlide(changed[0].index);
      }
    },
    [],
  );

  const viewabilityConfigCallbackPairs = useRef<ViewabilityConfigCallbackPairs>(
    [{ viewabilityConfig, onViewableItemsChanged }],
  );

  return (
    <View style={styles.container}>
      <FlatList
        data={homeSlides}
        horizontal
        showsHorizontalScrollIndicator={false}
        pagingEnabled
        renderItem={({ item }) => (
          <BannerItem item={item} width={layout.width}></BannerItem>
        )}
        numColumns={1}
        onLayout={(event: LayoutChangeEvent) =>
          setLayout(event.nativeEvent.layout)
        }
        viewabilityConfigCallbackPairs={viewabilityConfigCallbackPairs.current}
      />

      <View style={styles.indicatorWrapper}>
        {homeSlides.map((_, index) => {
          let activeStyle = {};
          if (index === currentActiveSlide) {
            activeStyle = { width: 32, backgroundColor: colors.primary.main };
          }

          return (
            <View
              testID="indicator"
              style={[styles.indicator, activeStyle]}
              key={index}
            />
          );
        })}
      </View>
    </View>
  );
};

export default Banners;
