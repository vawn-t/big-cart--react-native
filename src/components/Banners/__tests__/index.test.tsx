import Banners from '../index';

// Tests
import { render, screen } from '@utils/index';

// Mocks
import { homeSlides } from '@mocks/index';

// Themes
import { colors } from '@themes/index';

jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock'),
);

describe('Banners component', () => {
  const setup = () => {
    return render(<Banners />);
  };

  it('renders correctly', () => {
    const { toJSON } = setup();
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders the correct number of indicators', () => {
    setup();
    const indicators = screen.getAllByTestId('indicator');
    expect(indicators.length).toBe(homeSlides.length);
  });

  it('renders the first indicator as active', () => {
    setup();
    const indicators = screen.getAllByTestId('indicator');
    expect(indicators[0]).toHaveStyle({
      width: 32,
      backgroundColor: colors.primary.main,
    });
  });
});
