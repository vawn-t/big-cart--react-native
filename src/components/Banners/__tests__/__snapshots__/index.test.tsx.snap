// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banners component renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "height": 283,
              }
            }
          >
            <RCTScrollView
              data={
                [
                  {
                    "id": 1,
                    "image": 1,
                    "title": "20% off on your first purchase",
                  },
                  {
                    "id": 2,
                    "image": 1,
                    "title": "20% off on your first purchase",
                  },
                  {
                    "id": 3,
                    "image": 1,
                    "title": "20% off on your first purchase",
                  },
                  {
                    "id": 4,
                    "image": 1,
                    "title": "20% off on your first purchase",
                  },
                ]
              }
              getItem={[Function]}
              getItemCount={[Function]}
              horizontal={true}
              keyExtractor={[Function]}
              onContentSizeChange={[Function]}
              onLayout={[Function]}
              onMomentumScrollBegin={[Function]}
              onMomentumScrollEnd={[Function]}
              onScroll={[Function]}
              onScrollBeginDrag={[Function]}
              onScrollEndDrag={[Function]}
              pagingEnabled={true}
              removeClippedSubviews={false}
              renderItem={[Function]}
              scrollEventThrottle={50}
              showsHorizontalScrollIndicator={false}
              stickyHeaderIndices={[]}
              viewabilityConfigCallbackPairs={
                [
                  {
                    "onViewableItemsChanged": [Function],
                    "viewabilityConfig": {
                      "viewAreaCoveragePercentThreshold": 100,
                      "waitForInteraction": true,
                    },
                  },
                ]
              }
            >
              <View>
                <View
                  onFocusCapture={[Function]}
                  onLayout={[Function]}
                  style={
                    [
                      {
                        "flexDirection": "row",
                      },
                      null,
                    ]
                  }
                >
                  <View
                    accessibilityIgnoresInvertColors={true}
                    style={
                      [
                        {
                          "flex": 1,
                          "justifyContent": "center",
                          "paddingHorizontal": 45,
                          "paddingTop": 45,
                          "width": "100%",
                        },
                        {
                          "width": 0,
                        },
                      ]
                    }
                  >
                    <Image
                      resizeMode="cover"
                      source={1}
                      style={
                        [
                          {
                            "bottom": 0,
                            "left": 0,
                            "position": "absolute",
                            "right": 0,
                            "top": 0,
                          },
                          {
                            "height": undefined,
                            "width": 0,
                          },
                          undefined,
                        ]
                      }
                    />
                    <Text
                      style={
                        [
                          {
                            "color": "#000000",
                          },
                          {
                            "fontFamily": "Poppins-SemiBold",
                          },
                          {
                            "fontSize": 20,
                          },
                          {
                            "width": 175,
                          },
                        ]
                      }
                    >
                      20% off on your first purchase
                    </Text>
                  </View>
                </View>
                <View
                  onFocusCapture={[Function]}
                  onLayout={[Function]}
                  style={
                    [
                      {
                        "flexDirection": "row",
                      },
                      null,
                    ]
                  }
                >
                  <View
                    accessibilityIgnoresInvertColors={true}
                    style={
                      [
                        {
                          "flex": 1,
                          "justifyContent": "center",
                          "paddingHorizontal": 45,
                          "paddingTop": 45,
                          "width": "100%",
                        },
                        {
                          "width": 0,
                        },
                      ]
                    }
                  >
                    <Image
                      resizeMode="cover"
                      source={1}
                      style={
                        [
                          {
                            "bottom": 0,
                            "left": 0,
                            "position": "absolute",
                            "right": 0,
                            "top": 0,
                          },
                          {
                            "height": undefined,
                            "width": 0,
                          },
                          undefined,
                        ]
                      }
                    />
                    <Text
                      style={
                        [
                          {
                            "color": "#000000",
                          },
                          {
                            "fontFamily": "Poppins-SemiBold",
                          },
                          {
                            "fontSize": 20,
                          },
                          {
                            "width": 175,
                          },
                        ]
                      }
                    >
                      20% off on your first purchase
                    </Text>
                  </View>
                </View>
                <View
                  onFocusCapture={[Function]}
                  onLayout={[Function]}
                  style={
                    [
                      {
                        "flexDirection": "row",
                      },
                      null,
                    ]
                  }
                >
                  <View
                    accessibilityIgnoresInvertColors={true}
                    style={
                      [
                        {
                          "flex": 1,
                          "justifyContent": "center",
                          "paddingHorizontal": 45,
                          "paddingTop": 45,
                          "width": "100%",
                        },
                        {
                          "width": 0,
                        },
                      ]
                    }
                  >
                    <Image
                      resizeMode="cover"
                      source={1}
                      style={
                        [
                          {
                            "bottom": 0,
                            "left": 0,
                            "position": "absolute",
                            "right": 0,
                            "top": 0,
                          },
                          {
                            "height": undefined,
                            "width": 0,
                          },
                          undefined,
                        ]
                      }
                    />
                    <Text
                      style={
                        [
                          {
                            "color": "#000000",
                          },
                          {
                            "fontFamily": "Poppins-SemiBold",
                          },
                          {
                            "fontSize": 20,
                          },
                          {
                            "width": 175,
                          },
                        ]
                      }
                    >
                      20% off on your first purchase
                    </Text>
                  </View>
                </View>
                <View
                  onFocusCapture={[Function]}
                  onLayout={[Function]}
                  style={
                    [
                      {
                        "flexDirection": "row",
                      },
                      null,
                    ]
                  }
                >
                  <View
                    accessibilityIgnoresInvertColors={true}
                    style={
                      [
                        {
                          "flex": 1,
                          "justifyContent": "center",
                          "paddingHorizontal": 45,
                          "paddingTop": 45,
                          "width": "100%",
                        },
                        {
                          "width": 0,
                        },
                      ]
                    }
                  >
                    <Image
                      resizeMode="cover"
                      source={1}
                      style={
                        [
                          {
                            "bottom": 0,
                            "left": 0,
                            "position": "absolute",
                            "right": 0,
                            "top": 0,
                          },
                          {
                            "height": undefined,
                            "width": 0,
                          },
                          undefined,
                        ]
                      }
                    />
                    <Text
                      style={
                        [
                          {
                            "color": "#000000",
                          },
                          {
                            "fontFamily": "Poppins-SemiBold",
                          },
                          {
                            "fontSize": 20,
                          },
                          {
                            "width": 175,
                          },
                        ]
                      }
                    >
                      20% off on your first purchase
                    </Text>
                  </View>
                </View>
              </View>
            </RCTScrollView>
            <View
              style={
                {
                  "bottom": 30,
                  "columnGap": 5,
                  "flexDirection": "row",
                  "left": 15,
                  "position": "absolute",
                }
              }
            >
              <View
                style={
                  [
                    {
                      "backgroundColor": "#EBFFD7",
                      "borderRadius": 4,
                      "height": 8,
                      "width": 8,
                    },
                    {
                      "backgroundColor": "#AEDC81",
                      "width": 32,
                    },
                  ]
                }
                testID="indicator"
              />
              <View
                style={
                  [
                    {
                      "backgroundColor": "#EBFFD7",
                      "borderRadius": 4,
                      "height": 8,
                      "width": 8,
                    },
                    {},
                  ]
                }
                testID="indicator"
              />
              <View
                style={
                  [
                    {
                      "backgroundColor": "#EBFFD7",
                      "borderRadius": 4,
                      "height": 8,
                      "width": 8,
                    },
                    {},
                  ]
                }
                testID="indicator"
              />
              <View
                style={
                  [
                    {
                      "backgroundColor": "#EBFFD7",
                      "borderRadius": 4,
                      "height": 8,
                      "width": 8,
                    },
                    {},
                  ]
                }
                testID="indicator"
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="test"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
